### ISSUE TITLE
可以单独使用Qwen2.5-VL的Vision Encoder吗

### ISSUE THREAD

**Issue by Katherine015** [QwenLM/Qwen3-VL#1813]
https://github.com/QwenLM/Qwen3-VL/issues/1733

检视问题后发现问题一开始发生在

`hidden_states = hidden_states[window_index, :, :]`

参考的源码：
https://github.com/huggingface/transformers/blob/40a493c7ed4f19f08eadb0639cf26d49bfa5e180/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L377

呼叫的方式是：

`self.qwen = Qwen2_5_VLModel.from_pretrained( VLM_path['qwen2.5-vl-3B'], dtype=torch.float16, device_map="auto", trust_remote_code=True).eval()`

`img_token = self.qwen.get_image_features(img, grid_thw)`

其中img,shape=[B, 3, 504, 504], grid_thw计算方式是[1, H//14, W//14] * B 并转换成tensor，以504*504的影像大小，会得到grid_thw=tensor([[1, 36, 36]])
根据这个grid_thw以get_window_index计算得到的window_index数值会介于0~323之间：
`(tensor([  0,   1,   2,   3,  18,  19,  20,  21,  36,  37,  38,  39,  54,  55, 56,  57,   4,   5,   6,   7,  22,  23,  24,  25,  40,  41,  42,  43, 58,  59,  60,  61,   8,   9,  10,  11,  26,  27,  28,  29,  44,  45, 46,  47,  62,  63,  64,  65,  12,  13,  14,  15,  30,  31,  32,  33, 48,  49,  50,  51,  66,  67,  68,  69,  16,  17,  34,  35,  52,  53, 70,  71,  72,  73,  74,  75,  90,  91,  92,  93, 108, 109, 110, 111, 126, 127, 128, 129,  76,  77,  78,  79,  94,  95,  96,  97, 112, 113, 114, 115, 130, 131, 132, 133,  80,  81,  82,  83,  98,  99, 100, 101, 116, 117, 118, 119, 134, 135, 136, 137,  84,  85,  86,  87, 102, 103, 104, 105, 120, 121, 122, 123, 138, 139, 140, 141,  88,  89, 106, 107, 124, 125, 142, 143, 144, 145, 146, 147, 162, 163, 164, 165, 180, 181, 182, 183, 198, 199, 200, 201, 148, 149, 150, 151, 166, 167, 168, 169, 184, 185, 186, 187, 202, 203, 204, 205, 152, 153, 154, 155, 170, 171, 172, 173, 188, 189, 190, 191, 206, 207, 208, 209, 156, 157, 158, 159, 174, 175, 176, 177, 192, 193, 194, 195, 210, 211, 212, 213, 160, 161, 178, 179, 196, 197, 214, 215, 216, 217, 218, 219, 234, 235, 236, 237, 252, 253, 254, 255, 270, 271, 272, 273, 220, 221, 222, 223, 238, 239, 240, 241, 256, 257, 258, 259, 274, 275, 276, 277, 224, 225, 226, 227, 242, 243, 244, 245, 260, 261, 262, 263, 278, 279, 280, 281, 228, 229, 230, 231, 246, 247, 248, 249, 264, 265, 266, 267, 282, 283, 284, 285, 232, 233, 250, 251, 268, 269, 286, 287, 288, 289, 290, 291, 306, 307,308, 309, 292, 293, 294, 295, 310, 311, 312, 313, 296, 297, 298, 299, 314, 315, 316, 317, 300, 301, 302, 303, 318, 319, 320, 321, 304, 305, 322, 323])`

而在套用hidden_states = hidden_states[window_index, :, :]之前hidden_states的shape为[162, 4, 1280]

问题：
1. 请问以img_token = self.qwen.get_image_features(img, grid_thw)的过程套用vision encoder是正确的吗？
2. 如果套用方式是正確的，根據源碼觀察到的，影像輸入後會經過一個
`self.patch_embed = Qwen2_5_VisionPatchEmbed`
(https://github.com/huggingface/transformers/blob/40a493c7ed4f19f08eadb0639cf26d49bfa5e180/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L77)
hidden_state的shape会从输入的[1, 3, 504, 504]变成[648, 1280]，再经过
`hidden_states = hidden_states.reshape(seq_len // self.spatial_merge_unit, self.spatial_merge_unit, -1)`
后会变成[162, 4, 1280]，根据这样的结果，不论影像大小设定为多少，hidden_state的值(在这个情况下是162)永远无法满足window_index

麻烦您解答以上的困惑

**Comment by JJJYmmm** [comment]
For a image of shape [1, 3, 504, 504], after the 3d conv, the shape should be [1296, 1280], not [648,1280].

(3d conv's shape = (2, 14, 14), 504 / 14 = 36， 36*36 = 1296)

You should check the image preprocess logic.

**Comment by Katherine015** [comment]
> For a image of shape [1, 3, 504, 504], after the 3d conv, the shape should be [1296, 1280], not [648,1280].
> 
> (3d conv's shape = (2, 14, 14), 504 / 14 = 36， 36*36 = 1296)
> 
> You should check the image preprocess logic.

@JJJYmmm I checked  the caption of self.get_image_features, the image input shape is [B, 3, H, W]. So my input is a [1, 3, 504, 504] image tensor and the following step is the Qwen2_5_VisionPatchEmbed.
According to the source code (https://github.com/huggingface/transformers/blob/40a493c7ed4f19f08eadb0639cf26d49bfa5e180/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L77), 
`hidden_states = hidden_states.view(-1, self.in_channels, self.temporal_patch_size, self.patch_size, self.patch_size)`
will make the tensor shape become [648, 3, 2, 14, 14]. After that is the 3d conv and its shape will be [648, 1280, 1, 1].
Is the step wrong? Or the input shape of self.get_image_features is not [B, 3, H, W]?

**Comment by JJJYmmm** [comment]
Yes, you should preprocess the image. Here is an example.

 ```python
import torch
from transformers import AutoModelForImageTextToText, AutoProcessor


# 2 input images
images = torch.randn(2, 3, 504, 504)
grid_thw = torch.tensor([[1, 36, 36], [1, 36, 36]])
temporal_patch_size = 2
spatial_merge_size = 2
patch_size = 14

channel = images.shape[1]

pixel_values = []
for img, (grid_t, gird_h, gird_w) in zip(images, grid_thw):

    img = img.unsqueeze(0)
    # img shape [1, channel, H, W]

    # rpeat along the temporal dimension
    if img.shape[0] % temporal_patch_size != 0:
        repeats = img[-1].unsqueeze(0).repeat(temporal_patch_size - 1, 1, 1, 1)
        img = torch.cat([img, repeats], dim=0)

    # img shape [2, channel, H, W] for 3d conv

    img = img.view(grid_t, temporal_patch_size, channel, gird_h // spatial_merge_size, spatial_merge_size, patch_size, gird_w // spatial_merge_size, spatial_merge_size, patch_size)
    img = img.permute(0, 3, 6, 4, 7, 2, 1, 5, 8)
    flatten_patches = img.reshape(grid_t * gird_h * gird_w, channel * temporal_patch_size * patch_size * patch_size)
    pixel_values.append(flatten_patches)

pixel_values = torch.cat(pixel_values, dim=0)

print(pixel_values.shape)
```

**Comment by JJJYmmm** [comment]
For more details, you can refer to https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_vl/image_processing_qwen2_vl_fast.py

**Comment by Katherine015** [comment]
> Yes, you should preprocess the image. Here is an example.
> 
> import torch
> from transformers import AutoModelForImageTextToText, AutoProcessor
> 
> 
> # 2 input images
> images = torch.randn(2, 3, 504, 504)
> grid_thw = torch.tensor([[1, 36, 36], [1, 36, 36]])
> temporal_patch_size = 2
> spatial_merge_size = 2
> patch_size = 14
> 
> channel = images.shape[1]
> 
> pixel_values = []
> for img, (grid_t, gird_h, gird_w) in zip(images, grid_thw):
> 
>    img = img.unsqueeze(0)
>    # img shape [1, channel, H, W]
> 
>    # rpeat along the temporal dimension
>    if img.shape[0] % temporal_patch_size != 0:
>        repeats = img[-1].unsqueeze(0).repeat(temporal_patch_size - 1, 1, 1, 1)
>        img = torch.cat([img, repeats], dim=0)
> 
>    # img shape [2, channel, H, W] for 3d conv
> 
>    img = img.view(grid_t, temporal_patch_size, channel, gird_h // spatial_merge_size, spatial_merge_size, patch_size, gird_w // spatial_merge_size, spatial_merge_size, patch_size)
>    img = img.permute(0, 3, 6, 4, 7, 2, 1, 5, 8)
>    flatten_patches = img.reshape(grid_t * gird_h * gird_w, channel * temporal_patch_size * patch_size * patch_size)
>    pixel_values.append(flatten_patches)
> 
> pixel_values = torch.cat(pixel_values, dim=0)
> 
> print(pixel_values.shape)

The problem is solved. Thanks for your help
